chain/events/filter/index.go

```diff
@@ -67,34 +63,68 @@ var (
 )
 
 const (
-	eventExists          = `SELECT MAX(id) FROM event WHERE height=? AND tipset_key=? AND tipset_key_cid=? AND emitter_addr=? AND event_index=? AND message_cid=? AND message_index=?`
+	createTableEventsSeen = `CREATE TABLE IF NOT EXISTS events_seen (
+		id INTEGER PRIMARY KEY,
+		height INTEGER NOT NULL,
+		tipset_key_cid BLOB NOT NULL,
+		reverted INTEGER NOT NULL,
+	    UNIQUE(height, tipset_key_cid)
+	)`
+
 	insertEvent          = `INSERT OR IGNORE INTO event(height, tipset_key, tipset_key_cid, emitter_addr, event_index, message_cid, message_index, reverted) VALUES(?, ?, ?, ?, ?, ?, ?, ?)`
 	insertEntry          = `INSERT OR IGNORE INTO event_entry(event_id, indexed, flags, key, codec, value) VALUES(?, ?, ?, ?, ?, ?)`
 	revertEventsInTipset = `UPDATE event SET reverted=true WHERE height=? AND tipset_key=?`
 	restoreEvent         = `UPDATE event SET reverted=false WHERE height=? AND tipset_key=? AND tipset_key_cid=? AND emitter_addr=? AND event_index=? AND message_cid=? AND message_index=?`
 	revertEventSeen      = `UPDATE events_seen SET reverted=true WHERE height=? AND tipset_key_cid=?`
 	restoreEventSeen     = `UPDATE events_seen SET reverted=false WHERE height=? AND tipset_key_cid=?`
 	upsertEventsSeen     = `INSERT INTO events_seen(height, tipset_key_cid, reverted) VALUES(?, ?, false) ON CONFLICT(height, tipset_key_cid) DO UPDATE SET reverted=false`
-	isTipsetProcessed    = `SELECT COUNT(*) > 0 FROM events_seen WHERE tipset_key_cid=?`
-	getMaxHeightInIndex  = `SELECT MAX(height) FROM events_seen`
-	isHeightProcessed    = `SELECT COUNT(*) > 0 FROM events_seen WHERE height=?`
 
-	createIndexEventEmitterAddr  = `CREATE INDEX IF NOT EXISTS event_emitter_addr ON event (emitter_addr)`
+	eventExists = `SELECT MAX(id) FROM event WHERE height=? AND tipset_key=? AND tipset_key_cid=? AND emitter_addr=? AND event_index=? AND message_cid=? AND message_index=?`
+	// QUERY PLAN
+	// `--SEARCH event USING INDEX event_height (height=?)
+	isTipsetProcessed = `SELECT COUNT(*) > 0 FROM events_seen WHERE tipset_key_cid=?`
+	// QUERY PLAN
+	// `--SEARCH events_seen USING COVERING INDEX events_seen_tipset_key_cid (tipset_key_cid=?)
+	getMaxHeightInIndex = `SELECT MAX(height) FROM events_seen`
+	// QUERY PLAN
+	// `--SEARCH events_seen USING COVERING INDEX events_seen_height
+	isHeightProcessed = `SELECT COUNT(*) > 0 FROM events_seen WHERE height=?`
+	// QUERY PLAN
+	// `--SEARCH events_seen USING COVERING INDEX events_seen_height (height=?)
+
+	// When modifying indexes in this file, it is critical to test the query plan (EXPLAIN QUERY PLAN)
+	// of all the variations of queries built by prefillFilter to ensure that the query first hits
+	// an index that narrows down results to an epoch or a reasonable range of epochs. Specifically,
```

Comment:
Curious: is there a way to mechanize this?  Basically, can we iterate over the queries build by prefillFilter and assert that the output of EXPLAIN QUERY PLAN has event_tipset_key_cid or event_height should as the first index?

Reply:
I did start investigating this but didn't take it far enough to have a definitive answer; without full testing I believe this is a `sqlite3` client thing rather than something we can invoke programmatically, but I should check further.

Reply:
done in #12267